feat(assets): add account group balance selectors and components

amitabh94 opened the pull request:
## **Description**

Implement aggregated balance and portfolio change on the Wallet home screen (feature-flagged) using core calculators to mirror the extension behavior.

- Aggregated balance selectors in `app/selectors/assets/balances.ts` using core calculators and narrowed controller slices:
  - `selectBalanceForAllWallets`
  - `selectSelectedGroupAggregatedBalance`
  - `selectBalanceByAccountGroup`
  - `selectBalanceByWallet`
  - `selectBalanceChangeForAllWallets`
  - `selectBalanceChangeByAccountGroup`
  - `selectBalancePercentChange`
  - `selectBalancePercentChangeByAccountGroup`

- Narrowed state extractors for accounts, tokens, currency rates, and account tree (only fields required by calculators)

- UI wiring in `app/components/Views/Wallet/index.tsx`:
  - When `isMultichainState2Enabled` is true, render new `AccountGroupBalance` component 
  - Keep legacy rendering when the feature flag is off`

- Tests:
  - `app/selectors/assets/balances.test.ts` covering aggregate-by-wallet and zero/fallback cases
  - `app/components/UI/Assets/BalanceChange/AccountGroupBalanceChange.test.tsx` covering positive/negative changes and privacy mode




> **P.S. The aggregated balances are missing the native balances and staked balances as of now which will automatically be added in the next assets controller release.**


## **Changelog**

CHANGELOG entry: null

## **Related issues**

Fixes: N/A

## **Manual testing steps**

```gherkin
Feature: Aggregated balance and portfolio change (mobile)

  Scenario: Total balance and 1d change with feature flag enabled
    Given the feature flag "isMultichainState2Enabled" is enabled
    And the app is built with @metamask-previews/assets-controllers@73.1.0-preview-a89812f5
    And multiple accounts with tokens exist across chains
    When I open the Wallet home screen
    Then I see the total balance equal to the aggregated fiat value of all wallets
    And I see the 1d portfolio change row with amount and percent for the selected account group

  Scenario: Account group switch updates values
    Given the feature flag is enabled
    And I have at least two account groups with different holdings
    When I switch the selected account group
    Then the total balance and 1d change update to reflect the new group's aggregation only

  Scenario: Legacy behavior when feature flag disabled
    Given the feature flag "isMultichainState2Enabled" is disabled
    When I open the Wallet home screen
    Then the legacy balance and change components render instead of aggregated values

  Scenario: Privacy mode formatting parity
    Given privacy mode is enabled
    When the Wallet home screen renders
    Then the balance text uses the same obscured length and formatting as before

  Scenario: Empty state
    Given no token balances are available
    When the Wallet home screen renders
    Then the total balance displays 0 and portfolio change displays 0 amount and 0%
```

## **Screenshots/Recordings**

<div>
    <a href="https://www.loom.com/share/fcfc633b4b1b461fb48f92d9428c5476">
      <p>Aggregated Balances - BIP 44 - Watch Video</p>
    </a>
    <a href="https://www.loom.com/share/fcfc633b4b1b461fb48f92d9428c5476">
      <img style="max-width:300px;" src="https://cdn.loom.com/sessions/thumbnails/fcfc633b4b1b461fb48f92d9428c5476-7b981ea6e53a8a77-full-play.gif">
    </a>
  </div>

### **Before**


### **After**

## **Pre-merge author checklist**

- [x] I’ve followed [MetaMask Contributor Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Mobile Coding Standards](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've completed the PR template to the best of my ability
- [x] I’ve included tests if applicable
- [x] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [x] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

amitabh94 committed:
chore: update @metamask/assets-controllers to version 73.1.0

- Upgraded `@metamask/assets-controllers` to version `73.1.0` in both package.json and yarn.lock.

This update ensures compatibility with the latest features and improvements in the assets controllers.

amitabh94 committed:
feat(assets): Home uses aggregated balance for selected account group behind multichain flag; add temporary groupId derivation TODO

amitabh94 committed:
chore: remove obsolete assets-controllers patch after bump to ^73.1.0

amitabh94 committed:
chore(assets): clarify TODO on temporary groupId derivation; replace with selected account group id selector when available

amitabh94 committed:
refactor(assets): remove React.memo and dev logs; simplify groupId resolver to membership-first

amitabh94 committed:
refactor(assets): reduce React.useMemo usage in PortfolioBalance; keep selector factory memo only

amitabh94 committed:
refactor(assets): remove outdated TODO/comments and placeholders from PortfolioBalance

amitabh94 committed:
chore(assets): update TODO documenting temporary groupId resolver and long-term selector plan

amitabh94 committed:
style(assets): use named useMemo import for selector factory memoization

amitabh94 committed:
refactor(assets): remove hasAggregatedGroup; gate aggregated display on resolvedGroupBalance with flag

amitabh94 committed:
chore(assets): remove stale comment above selector factory

amitabh94 committed:
feat(assets): implement aggregated portfolio change component and selectors for balance changes

amitabh94 committed:
chore(assets): update assets-controllers dependency to npm package version 73.1.0-preview-343f3d69

amitabh94 committed:
chore(dependencies): update @metamask/base-controller to version 8.1.0 and consolidate version entries in yarn.lock

amitabh94 committed:
feat(assets): add AccountGroupBalanceChange component for displaying balance changes

amitabh94 committed:
chore(CODEOWNERS): add Assets components to CODEOWNERS for consolidating asset specific components

amitabh94 committed:
chore: restore patches/@metamask+assets-controllers+73.0.2.patch for compatibility

amitabh94 committed:
chore(assets): update patch file on preview build

amitabh94 committed:
test(assets): enhance tests for PortfolioBalance and asset selectors with mocked data and feature flags

amitabh94 committed:
Merge branch 'main' into feat/aggregated-balances-portfolio-change

amitabh94 committed:
chore: update @metamask/assets-controllers and @metamask/controller-utils dependencies

- Updated @metamask/assets-controllers to version 73.1.0-preview-a89812f5.
- Updated @metamask/controller-utils to version 11.12.0 in package.json and yarn.lock.

amitabh94 committed:
test: enhance PortfolioBalance tests with feature flag and balance selector mocks

- Updated tests for PortfolioBalance to mock feature flag behavior and balance selectors.
- Added scenarios for both legacy and new aggregated balance displays based on feature flag state.
- Improved error handling tests for selector errors and missing account tree data.
- Refactored balance selector mocks for better control and clarity in tests.

amitabh94 committed:
test: add comprehensive tests for AccountGroupBalanceChange component

- Introduced a new test suite for the AccountGroupBalanceChange component to validate rendering of positive, negative, and zero balance changes.
- Added tests for various scenarios including privacy mode, different currencies, and edge cases like large, small, infinity, and NaN values.
- Removed outdated tests and styles related to AggregatedChangeAccountGroup to streamline the codebase.

amitabh94 committed:
refactor: remove unused balance selectors and related tests

- Eliminated `selectSelectedGroupAggregatedBalance` and its associated tests to streamline the codebase.
- Updated tests to focus on `selectBalanceByAccountGroup` and removed redundant balance percent change selectors.
- This refactor enhances maintainability by reducing complexity in the asset balance selectors.

amitabh94 committed:
chore: update @metamask/assets-controllers to version 73.2.0

- Upgraded the @metamask/assets-controllers dependency in package.json and yarn.lock to version 73.2.0 for improved functionality and performance.

amitabh94 committed:
feat: implement AccountGroupBalance and AccountGroupBalanceChange components

- Added AccountGroupBalance and AccountGroupBalanceChange components to display account group balances and changes.
- Integrated balance selectors to fetch and format balance data based on the selected account group.
- Introduced tests for AccountGroupBalanceChange to validate rendering of various balance change scenarios, including privacy mode and different currencies.
- Updated PortfolioBalance to conditionally render AccountGroupBalance based on feature flag state.
- Enhanced balance selectors to support new functionality and improve maintainability.

amitabh94 committed:
Merge branch 'main' into feat/aggregated-balances-portfolio-change

amitabh94 committed:
chore: update @metamask/assets-controllers to version 73.3.0

amitabh94 committed:
Merge branch 'main' into feat/aggregated-balances-portfolio-change

amitabh94 committed:
Merge branch 'main' into feat/aggregated-balances-portfolio-change

amitabh94 committed:
chore: remove deduplicated @metamask/keyring-api dependency from yarn.lock

amitabh94 committed:
Merge branch 'main' into feat/aggregated-balances-portfolio-change

amitabh94 committed:
chore(patches): restore @metamask/assets-controllers+73.3.0.patch

amitabh94 committed:
chore: remove AggregatedPortfolioChange component

This commit deletes the AggregatedPortfolioChange component from the codebase, as it is no longer needed. The removal helps streamline the component library and reduce unnecessary code.

amitabh94 committed:
refactor(tests): simplify PortfolioBalance test setup by removing unused mocks

This commit refactors the PortfolioBalance test file by removing unnecessary mock implementations and related code that are no longer needed. This streamlining enhances test clarity and maintainability.

amitabh94 committed:
refactor: update styles and structure for improved layout

This commit refactors the AccountGroupBalance component by replacing the previous style implementation with a new approach using React Native's StyleSheet. The updated styles enhance the layout and organization of the component, including the addition of a skeleton loader for better user experience during data loading. This change streamlines the component's structure and improves maintainability.

amitabh94 committed:
Merge branch 'main' into feat/aggregated-balances-portfolio-change

amitabh94 committed:
refactor: simplify AccountGroupBalance component structure

This commit refactors the AccountGroupBalance component by removing unnecessary props from the View element, streamlining the component's structure for improved readability and maintainability. The change enhances the overall layout without altering the component's functionality.

amitabh94 committed:
refactor(tests): enhance AccountCell tests with configurable mock balance

This commit improves the AccountCell test suite by introducing a configurable mock balance for the balance selector, allowing for more flexible and accurate testing of balance display. The tests now validate the correct formatting of balances in different currencies, enhancing test coverage and maintainability.

amitabh94 committed:
Merge branch 'main' into feat/aggregated-balances-portfolio-change

amitabh94 committed:
Merge branch 'main' into feat/aggregated-balances-portfolio-change

amitabh94 committed:
test: Refactor `AccountListCell` tests to use `renderWithProvider` and mock balance selector

- Updated tests for `AccountListCell` to utilize `renderWithProvider` for improved context handling.
- Introduced a configurable mock balance to reduce dependencies on the store.
- Enhanced the balance selector mock to return a consistent mock balance for testing.

amitabh94 committed:
Merge branch 'main' into feat/aggregated-balances-portfolio-change

amitabh94 committed:
test: Update `AccountGroupBalanceChange` tests for formatting and visibility checks

- Adjusted tests to verify the correct formatting of percentage values, ensuring leading plus signs are included for positive changes.
- Updated visibility checks for hidden values to align with portfolio specifications, confirming the use of 10 dots for obscured amounts.
- Enhanced currency rendering tests to validate symbol-first formatting for USD and EUR, and code-suffix for GBP with trailing space.
- Simplified class presence checks to ensure elements are defined without relying on className, improving test reliability.

amitabh94 committed:
test: Enhance mock state in `createMockState` for multichain account testing

- Added minimal controllers required by balance selectors used in `AccountCell` to ensure accurate testing.
- Included safe default values for optional controllers to prevent downstream selector assumptions.
- Updated the mock state structure to support multichain balances and token management, improving test reliability and coverage.

amitabh94 committed:
test: Enhance mock state for `AccountGroupBalance` and `createMockState`

- Integrated `backgroundState` into `createMockState` to improve multichain account testing reliability.
- Updated `AccountGroupBalance` tests to utilize a more comprehensive mock state, ensuring accurate rendering and balance checks.
- Removed unnecessary minimal controllers from the mock state to streamline testing and prevent selector assumptions.

amitabh94 committed:
refactor: Remove `privacyMode` prop from `AccountGroupBalanceChange` and utilize Redux state

- Eliminated the `privacyMode` prop from `AccountGroupBalanceChange` to streamline component usage.
- Integrated `useSelector` to access `privacyMode` directly from Redux state, enhancing state management.
- Updated tests to reflect the removal of the prop and ensure accurate rendering based on privacy mode state.

amitabh94 committed:
test: Add selectors for balance and balance change by selected account group

amitabh94 committed:
Merge branch 'main' into feat/aggregated-balances-portfolio-change

amitabh94 committed:
chore: clean up CODEOWNERS file by removing unnecessary blank line

gantunesr commented on app/components/UI/Assets/components/Balance/AccountGroupBalance.tsx:
```diff
@@ -0,0 +1,87 @@
+    [PreferencesController],
+  );
+
+  const displayBalance = (() => {
```
Should this be wrapped in a `useCallback` hook?

amitabh94 replied:
Moved the formatting to within the selector itself

sethkfman reviewed (dismissed):
LGTM - Reviewed Wallet & CODEOWNER file

gantunesr commented on app/components/UI/Assets/components/Balance/AccountGroupBalance.tsx:
```diff
@@ -0,0 +1,87 @@
+          </TouchableOpacity>
+        ) : (
+          <View style={styles.skeletonContainer}>
+            <Skeleton width={100} height={40} />
+            <Skeleton width={100} height={20} />
```
Should this be part of a styles file?

gantunesr replied:
I thought it was the pattern we were following, if it is not the case then nevermind

tommasini commented on app/selectors/assets/balances.ts:
```diff
@@ -0,0 +1,348 @@
+    currencyRateState: CurrencyRateState,
+    enabledNetworkMap: Record<string, Record<string, boolean>> | undefined,
+  ) =>
+    calculateBalanceForAllWallets(
```
I believe as a future performance improvement, this can be used in worklets 
https://github.com/margelo/react-native-worklets-core

I'm saying this because I believe that the time complexity and space complexity of the [calculateBalancesForWallets](https://github.com/MetaMask/core/blob/b7b8be4608b86a969996d4e3bcc21c4cc25c2de3/packages/assets-controllers/src/balances.ts#L301) function could have performance challenges (FPS drop), at Android lower end devices

We should double check the behavior there with a power user account

Prithpal-Sooriya replied:
@tommasini really nice callout, this is a great use-case!

owencraston commented on app/component-library/components-temp/MultichainAccounts/AccountCell/AccountCell.tsx:
```diff
@@ -30,6 +34,23 @@ const AccountCell = ({ accountGroup, isSelected }: AccountCellProps) => {
+    if (!groupBalance) {
+      return undefined;
+    }
+    const value = groupBalance.totalBalanceInUserCurrency;
```
why not move the formatting logic inside the selector. this mimics a backend/front end model.

amitabh94 replied:
On second thought , I have decided to keep it out as formatting in selectors adds sync CPU on every recompute and creates new strings/objects, breaking memoization and causing extra re-renders. We’ll keep selectors numeric and format at the component edge (memoized) to keep lists and the JS thread responsive.

amitabh94 committed:
test: Update balance selectors to include formatted balance

- Enhanced the `selectBalanceByAccountGroup` and `selectBalanceBySelectedAccountGroup` selectors to return a formatted balance string.
- Updated `AccountCell` and `AccountGroupBalance` components to utilize the new formatted balance property, improving display consistency.
- Adjusted tests for `AccountCell` and `AccountGroupBalance` to validate the rendering of formatted balances, ensuring accurate UI representation.

amitabh94 committed:
Merge branch 'main' into feat/aggregated-balances-portfolio-change

amitabh94 committed:
test: Refactor AccountGroupBalance tests for formatted balance rendering

- Updated tests for `AccountGroupBalance` to assert the rendering of formatted balance strings using a direct text query.
- Improved clarity and consistency in test assertions by utilizing `getByText` for balance verification.

amitabh94 committed:
Revert "test: Refactor AccountGroupBalance tests for formatted balance rendering"

This reverts commit 40f72981578e7a9c56bbefa7a7cea0ea9e3f0909.

amitabh94 committed:
Revert "test: Update balance selectors to include formatted balance"

This reverts commit 90cdd520e3ac6e4231f4fbe7ebc3147a50827c3f.

amitabh94 committed:
refactor: optimize balance display logic in Account components

- Updated `AccountCell`, `AccountGroupBalance`, and `AccountGroupBalanceChange` components to utilize `useMemo` for calculating display balances, improving performance by preventing unnecessary re-renders.
- Simplified balance retrieval by directly accessing `totalBalanceInUserCurrency` and `userCurrency` from the `groupBalance` state.
- Enhanced clarity in balance formatting by ensuring proper checks for null values before rendering.

amitabh94 committed:
Merge branch 'main' into feat/aggregated-balances-portfolio-change

amitabh94 committed:
Merge branch 'main' into feat/aggregated-balances-portfolio-change

gantunesr commented on app/component-library/components-temp/MultichainAccounts/AccountCell/AccountCell.tsx:
```diff
@@ -30,6 +34,23 @@ const AccountCell = ({ accountGroup, isSelected }: AccountCellProps) => {
     });
   }, [navigate, accountGroup]);
 
+  const selectBalanceForGroup = useMemo(
+    () => selectBalanceByAccountGroup(accountGroup.id),
+    [accountGroup.id],
+  );
```
Final decision is to refactor it in a follow up PR
